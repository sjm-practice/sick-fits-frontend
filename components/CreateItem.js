import React, { Component } from "react";
import { Mutation } from "react-apollo";
import Router from "next/router";
import gql from "graphql-tag";
import Form from "./styles/Form";
import ErrorMessage from "./ErrorMessage";

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: "Sample",
    description: "Sample Description",
    image: "",
    largeImage: "large-sample.jpg", // eslint-disable-line react/no-unused-state
    price: 1000,
  };

  handleChange = e => {
    const { name, type, value } = e.target;
    const val = type === "number" ? parseFloat(value) : value;

    this.setState({ [name]: val });
  };

  handleSubmit = async (e, createItem) => {
    e.preventDefault();
    const res = await createItem();
    console.log(res);
    Router.push({
      pathname: "/item",
      query: { id: res.data.createItem.id },
    });
  };

  uploadFile = async e => {
    const { files } = e.target;
    const data = new FormData();
    data.append("file", files[0]);
    data.append("upload_preset", "sickfits");

    const res = await fetch("https://api.cloudinary.com/v1_1/smarsh/image/upload", {
      method: "POST",
      body: data,
    });
    const file = await res.json();
    console.log(file);
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url, // eslint-disable-line react/no-unused-state
    });
  };

  render() {
    const { title, price, description, image } = this.state;

    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form onSubmit={e => this.handleSubmit(e, createItem)}>
            <ErrorMessage error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="image">
                Image
                <input
                  type="file"
                  id="image"
                  name="image"
                  required
                  placeholder="Upload An Image"
                  onChange={this.uploadFile}
                />
                {image && <img width="200" src={image} alt="Upload Preview" />}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  required
                  value={price}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter A Description"
                  required
                  value={description}
                  onChange={this.handleChange}
                />
              </label>
            </fieldset>
            <button type="submit">Submit</button>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
