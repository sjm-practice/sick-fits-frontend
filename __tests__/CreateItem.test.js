import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Router from "next/router";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import CreateItem, { CREATE_ITEM_MUTATION } from "../components/CreateItem";
import { fakeItem } from "../lib/testUtils";

const dogImage = "https://dogs.com/dog.jpg";

// mock global fetch api
global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: dogImage,
    eager: [{ secure_url: dogImage }],
  }),
});

describe("<CreateItem />", () => {
  afterAll(() => global.fetch.mockReset());

  it("should render and match snapshot", async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>,
    );

    expect(toJSON(wrapper.find("form[data-test='form']"))).toMatchSnapshot();
  });

  it("should upload a file when file input changes", async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>,
    );

    const input = wrapper.find("input[type='file']");
    input.simulate("change", { target: { files: ["dogimage.jpg"] } });
    await wait();

    // check the state of the component instance
    const component = wrapper.find("CreateItem").instance();
    expect(component.state.image).toBe(dogImage);
    expect(component.state.largeImage).toBe(dogImage);
    expect(global.fetch).toHaveBeenCalled();
  });

  it("should save form values in state", async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>,
    );

    wrapper
      .find("#title")
      .simulate("change", { target: { name: "title", value: "Title" } });
    wrapper
      .find("#price")
      .simulate("change", { target: { name: "price", value: 500, type: "number" } });
    wrapper
      .find("#description")
      .simulate("change", { target: { name: "description", value: "Description" } });

    await wait();

    expect(wrapper.find("CreateItem").instance().state).toMatchObject({
      title: "Title",
      price: 500,
      description: "Description",
    });
  });

  it("should create an item when the form is submitted", async () => {
    const item = fakeItem();
    const mocks = [
      {
        request: {
          query: CREATE_ITEM_MUTATION,
          variables: {
            title: item.title,
            price: item.price,
            description: item.description,
            image: "",
            largeImage: "",
          },
        },
        result: {
          data: {
            createItem: {
              ...item,
              __typeName: "Item",
            },
          },
        },
      },
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CreateItem />
      </MockedProvider>,
    );

    // simulate user filling out the form
    wrapper
      .find("#title")
      .simulate("change", { target: { name: "title", value: item.title } });
    wrapper.find("#price").simulate("change", {
      target: { name: "price", value: item.price, type: "number" },
    });
    wrapper
      .find("#description")
      .simulate("change", { target: { name: "description", value: item.description } });

    // mock the router
    Router.router = { push: jest.fn() };

    wrapper.find("form").simulate("submit");
    await wait(50);

    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: "/item",
      query: { id: item.id },
    });
  });
});
