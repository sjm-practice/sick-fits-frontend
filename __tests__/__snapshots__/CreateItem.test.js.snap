// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> should render and match snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="image"
    >
      Image
      <input
        id="image"
        name="image"
        onChange={[Function]}
        placeholder="Upload An Image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter A Description"
        required={true}
        value=""
      />
    </label>
  </fieldset>
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
